File: clients/js/src/signed_calls.ts
import { BlockTag } from '@ethersproject/abstract-provider';
import {
  Signer,
  TypedDataDomain,
  TypedDataField,
  TypedDataSigner,
} from '@ethersproject/abstract-signer';
import { BigNumber, BigNumberish } from '@ethersproject/bignumber';
import { BytesLike, arrayify, hexlify } from '@ethersproject/bytes';
import * as cbor from 'cborg';
import type {
  CamelCasedProperties,
  Promisable,
  RequireExactlyOne,
  SetRequired,
} from 'type-fest';

const DEFAULT_GAS_PRICE = 0;
const DEFAULT_GAS_LIMIT = 0;
const DEFAULT_VALUE = 0;
const DEFAULT_DATA = '0x';
const zeroAddress = () => `0x${'0'.repeat(40)}`;

export function signedCallEIP712Params(chainId: number): {
  domain: TypedDataDomain;
  types: Record<string, TypedDataField[]>;
} {
  return {
    domain: {
      name: 'oasis-runtime-sdk/evm: signed query',
      version: '1.0.0',
      chainId,
    },
    types: {
      Call: [
        { name: 'from', type: 'address' },
        { name: 'to', type: 'address' },
        { name: 'gasLimit', type: 'uint64' },
        { name: 'gasPrice', type: 'uint256' },
        { name: 'value', type: 'uint256' },
        { name: 'data', type: 'bytes' },
        { name: 'leash', type: 'Leash' },
      ],
      Leash: [
        { name: 'nonce', type: 'uint64' },
        { name: 'blockNumber', type: 'uint64' },
        { name: 'blockHash', type: 'bytes32' },
        { name: 'blockRange', type: 'uint64' },
      ],
    },
  };
}

/** Prepares a signed call that allows retrieving data scoped to the sender's account. */
export async function prepareSignedCall(
  call: EthCall,
  signer: Signer & TypedDataSigner,
  overrides?: PrepareSignedCallOverrides,
): Promise<SignedCall> {
  if (!call.from) {
    throw TypeError('signed call must have a sender');
  }
  const leash = await makeLeash(signer, overrides?.leash);
  const extPack: SignedCallDataPack = {
    data: call.data ? { body: arrayify(call.data) } : undefined,
    leash,
    signature: await signCall(makeSignableCall(call, leash), signer, {
      chainId: overrides?.chainId,
    }),
  };
  return {
    ...call,
    data: hexlify(cbor.encode(extPack)),
  };
}

async function makeLeash(
  signer: Signer & TypedDataSigner,
  overrides?: LeashOverrides,
): Promise<Leash> {
  const nonceP = overrides?.nonce
    ? overrides.nonce
    : signer.getTransactionCount('pending');
  let blockP: Promisable<BlockId>;
  if (overrides?.block !== undefined) {
    blockP = overrides.block;
  } else {
    if (signer.provider === undefined) {
      throw new Error(
        'unable to locate base block, as a provider is not connected',
      );
    }
    const blockTag = overrides?.blockTag ?? 'latest';
    blockP = signer.provider.getBlock(blockTag);
  }
  const [nonce, block] = await Promise.all([nonceP, blockP]);
  return {
    nonce,
    block_number: block.number,
    block_hash: arrayify(block.hash),
    block_range: overrides?.blockRange ?? 15 /* ~90s */,
  };
}

export function makeSignableCall(call: EthCall, leash: Leash): SignableEthCall {
  return {
    from: call.from,
    to: call.to ?? zeroAddress(),
    gasLimit: BigNumber.from(
      call.gas ?? call.gasLimit ?? DEFAULT_GAS_LIMIT,
    ).toNumber(),
    gasPrice: BigNumber.from(call.gasPrice ?? DEFAULT_GAS_PRICE),
    value: BigNumber.from(call.value ?? DEFAULT_VALUE),
    data: call.data
      ? hexlify(call.data, { allowMissingPrefix: true })
      : DEFAULT_DATA,
    leash: {
      nonce: leash.nonce,
      blockNumber: leash.block_number,
      blockHash: leash.block_hash,
      blockRange: leash.block_range,
    },
  };
}

async function signCall(
  call: SignableEthCall,
  signer: Signer & TypedDataSigner,
  overrides?: Partial<{ chainId: number }>,
): Promise<Uint8Array> {
  const chainId = overrides?.chainId ?? (await signer.getChainId());
  if (!overrides?.chainId && chainId !== 0x5afe && chainId !== 0x5aff) {
    throw new Error(
      'Signed queries can only be sent to Sapphire or Sapphire Testnet. Please check your Web3 provider connection.',
    );
  }
  const { domain, types } = signedCallEIP712Params(chainId);
  return arrayify(await signer._signTypedData(domain, types, call));
}

export type PrepareSignedCallOverrides = Partial<{
  leash: LeashOverrides;
  chainId: number;
}>;

export type LeashOverrides = Partial<
  {
    nonce: number;
    blockRange: number;
  } & RequireExactlyOne<{
    block: BlockId;
    blockTag: BlockTag;
  }>
>;

export type EthCall = {
  /** 0x-prefixed hex-encoded address. */
  from: string;
  /** Optional 0x-prefixed hex-encoded address. */
  to?: string;
  value?: BigNumberish;
  gasPrice?: BigNumberish;
  data?: BytesLike;
} & Partial<
  RequireExactlyOne<{
    gas: number | string; // web3.js
    gasLimit: BigNumberish; // ethers
  }>
>;

export type SignedCall = SetRequired<EthCall, 'data'>;

/**
 * Parameters that define a signed call that shall be
 * CBOR-encoded and sent as the call's `data` field.
 */
export type SignedCallDataPack = {
  /**
   * An oasis-sdk `Call` without the optional fields.
   *
   * After encryption, `body` would be encryped and this field would contain a
   * `format` field. The runtime would decode the data as a `types::transaction::Call`.
   **/
  data?: { body: Uint8Array };
  leash: Leash;
  /** A signature over the call and leash as generated by `signCall`. */
  signature: Uint8Array;
};

/**
 * The structure passed to eth_signTypedData_v4.
 *
 * `uint256`, `address`, and `bytes` are required to be hex-stringified.
 */
export type SignableEthCall = {
  from: string;
  to: string;
  gasLimit?: number;
  gasPrice?: BigNumber;
  value?: BigNumber;
  data?: string;
  leash: CamelCasedProperties<Leash>;
};

export type Leash = {
  /** The largest sender account nonce whence the call will be valid. */
  nonce: number;
  /** The block number whence the call will be valid. */
  block_number: number; // uint64
  /** The expected block hash to be found at `block_number`. */
  block_hash: Uint8Array;
  /** The number of blocks past the block at `block_number` whence the call will be valid. */
  block_range: number; // uint64
};

type BlockId = { hash: string; number: number };

File: clients/go/signed_calls.go
package sapphire

import (
	"encoding/hex"
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/common/math"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/signer/core/apitypes"
	"github.com/oasisprotocol/oasis-core/go/common/cbor"
)

const ZeroAddress = "0x0000000000000000000000000000000000000000"

// SignedCallDataPack defines a signed call.
//
// It should be encoded and sent in the `data` field of an Ethereum call.
type SignedCallDataPack struct {
	Data      Data   `json:"data"`
	Leash     Leash  `json:"leash"`
	Signature []byte `json:"signature"`
}

// Data is the plain data in the datapack.
type Data struct {
	Body []byte `json:"body"`
}

// DataEnvelope is an oasis-sdk `Call` without optional fields.
type DataEnvelope struct {
	Body   []byte `json:"body"`
	Format uint64 `json:"format,omitempty"` // reuse for now, TODO swap later
}

// EncryptedBodyEnvelope is an oasis-sdk `Call` with optional fields.
type EncryptedBodyEnvelope struct {
	Body   Body   `json:"body"`
	Format uint64 `json:"format"`
}

type Body struct {
	PK    []byte `json:"pk"`
	Data  []byte `json:"data"`
	Nonce []byte `json:"nonce"`
}

type Leash struct {
	Nonce       uint64 `json:"nonce"`
	BlockNumber uint64 `json:"block_number"`
	BlockHash   []byte `json:"block_hash"`
	BlockRange  uint64 `json:"block_range"`
}

// NewDataPack returns a SignedCallDataPack.
//
// This method does not encrypt `data`, so that should be done afterwards.
func NewDataPack(sign SignerFn, chainID uint64, caller, callee []byte, gasLimit uint64, gasPrice, value *big.Int, data []byte, leash Leash) (*SignedCallDataPack, error) {
	signable := makeSignableCall(chainID, caller, callee, gasLimit, gasPrice, value, data, leash)
	signature, err := signTypedData(sign, signable)
	if err != nil {
		return nil, fmt.Errorf("failed to sign call: %w", err)
	}
	return &SignedCallDataPack{
		Data:      Data{Body: data},
		Leash:     leash,
		Signature: signature,
	}, nil
}

func (p SignedCallDataPack) Encode() []byte {
	return hexutil.Bytes(cbor.Marshal(p.Data.Body))
}

func (p SignedCallDataPack) EncryptEncode(cipher Cipher) []byte {
	// Encrypt when data exists
	if p.Data.Body != nil {
		return cipher.EncryptEncode(p.Data.Body)
	}

	return p.Encode()
}

func NewLeash(nonce uint64, blockNumber uint64, blockHash []byte, blockRange uint64) Leash {
	return Leash{
		Nonce:       nonce,
		BlockNumber: blockNumber,
		BlockHash:   blockHash,
		BlockRange:  blockRange,
	}
}

func makeSignableCall(chainID uint64, caller, callee []byte, gasLimit uint64, gasPrice *big.Int, value *big.Int, data []byte, leash Leash) apitypes.TypedData {
	toAddr := ZeroAddress
	// callee should exist except for contract creation
	if callee != nil {
		toAddr = hex.EncodeToString(callee[:])
	}

	if value == nil {
		value = big.NewInt(0)
	}
	valueU256 := math.HexOrDecimal256(*value)

	if gasPrice == nil {
		gasPrice = big.NewInt(0)
	}
	gasPriceU256 := math.HexOrDecimal256(*gasPrice)

	return apitypes.TypedData{
		Types: map[string][]apitypes.Type{
			"EIP712Domain": {
				{Name: "name", Type: "string"},
				{Name: "version", Type: "string"},
				{Name: "chainId", Type: "uint256"},
			},
			"Call": {
				{Name: "from", Type: "address"},
				{Name: "to", Type: "address"},
				{Name: "gasLimit", Type: "uint64"},
				{Name: "gasPrice", Type: "uint256"},
				{Name: "value", Type: "uint256"},
				{Name: "data", Type: "bytes"},
				{Name: "leash", Type: "Leash"},
			},
			"Leash": {
				{Name: "nonce", Type: "uint64"},
				{Name: "blockNumber", Type: "uint64"},
				{Name: "blockHash", Type: "bytes32"},
				{Name: "blockRange", Type: "uint64"},
			},
		},
		PrimaryType: "Call",
		Domain: apitypes.TypedDataDomain{
			Name:              "oasis-runtime-sdk/evm: signed query",
			Version:           "1.0.0",
			ChainId:           math.NewHexOrDecimal256(int64(chainID)),
			VerifyingContract: "",
			Salt:              "",
		},
		Message: map[string]interface{}{
			"from":     hex.EncodeToString(caller[:]),
			"to":       toAddr,
			"value":    &valueU256,
			"gasLimit": math.NewHexOrDecimal256(int64(gasLimit)),
			"gasPrice": &gasPriceU256,
			"data":     data,
			"leash": map[string]interface{}{
				"nonce":       math.NewHexOrDecimal256(int64(leash.Nonce)),
				"blockNumber": math.NewHexOrDecimal256(int64(leash.BlockNumber)),
				"blockHash":   leash.BlockHash,
				"blockRange":  math.NewHexOrDecimal256(int64(leash.BlockRange)),
			},
		},
	}
}

// signTypedData is based on go-ethereum/core/signer but modified to use an in-memory signer.
func signTypedData(sign SignerFn, typedData apitypes.TypedData) ([]byte, error) {
	domainSeparator, err := typedData.HashStruct("EIP712Domain", typedData.Domain.Map())
	if err != nil {
		return nil, fmt.Errorf("failed to hash EIP721Domain: %w", err)
	}
	typedDataHash, err := typedData.HashStruct(typedData.PrimaryType, typedData.Message)
	if err != nil {
		return nil, fmt.Errorf("failed to hash typed data: %w", err)
	}
	rawData := []byte(fmt.Sprintf("\x19\x01%s%s", string(domainSeparator), string(typedDataHash)))
	digest := crypto.Keccak256Hash(rawData)
	signature, err := sign(*(*[32]byte)(digest.Bytes()))
	if err != nil {
		return nil, fmt.Errorf("failed to sign typed data: %w", err)
	}
	signature[64] = 28 // Eth wallets use a high recovery ID.
	return signature, nil
}
